localstack/localstack
# Migrate DNS server to community

## Motivation

We want to simplify network configuration. A point of confusion is that `localhost.localstack.cloud` is publically registered, and resolves to `127.0.0.1`. When communicating between docker containers, this is an issue - trying to connect to e.g. an opensearch domain with url `*.localhost.localstack.cloud` returned by LocalStack will attempt to connect to the "application" container, rather than LocalStack.

Recently we have upgraded the DNS server provided by LocalStack Pro to dynamically resolve the IP address of the LocalStack container based on request subnet, but this functionality was only available in Pro.

With this PR we bring the DNS server from Pro to LocalStack community. This change brings along with it increased test coverage. 




## Changes

* Move the DNS server to community
* Add bootstrap tests covering the DNS server behaviour
* Start the DNS server regardless of whether the host port can be bound
* If port 53 (or `DNS_PORT`) can be bound on the host, publish the docker port using the CLI
* Move the DNS server module from `localstack.services.dns_server` to `localstack.dns.server` to group core resources (e.g. `http`, `dns`, etc.) at the top of the tree
* Remove broken attempts to configure the DNS of the host


## Testing

* Start LocalStack: `DNS_ADDRESS=127.0.0.1 localstack start`
* If port 53/udp is available on the host, then run `dig @127.0.0.1 localhost.localstack.cloud` and a valid response is returned
* Start a container in the same docker network/subnet as the LocalStack container, then run from that container `dig @<ip of localstack container> localhost.localstack.cloud` and the IP address of the container should be returned
* Remove broken attempts to configure the host DNS

## TODO

What's left to do:

- [x] Get -ext PR ready to merge
